packages/graphql/lib/src/links/websocket_link/websocket_client.dart

```diff
@@ -234,7 +247,9 @@ class SocketClient {
       _messageSubscription = _messages.listen(
         onMessage,
         onDone: onConnectionLost,
-        cancelOnError: true,
+        // onDone will not be triggered if the subscription is
+        // auto-cancelled on error; make sure to pass false
+        cancelOnError: false,
```

Comment:
For now, I have no in mind what is the use case that this change can not be wrong, but call `onDone` after an error sound wired to me.

can you explain what is the reason here?

Reply:
Right now the reconnection logic happens in the `onDone` callback (within `onConnectionLost`) and so if `onDone` is not called then nothing ever gets notified there is an end to the websocket connection, and so no reconnection attempt is made ever again. In this context, it seems `onDone` just refers to the end of a stream, and in dart a stream can receive multiple errors without the stream being done (in this particular stream it seems that it will normally only get one error). 

said another way, without this change, here is the current behavior:
* a websocket connection is closed by the server
* onStreamError callback is called (which looks to normally be set to _defaultOnStreamError), which just logs the exception (`WebSocketChannelException`, but doesn't do anything else), and then the stream subscription is automatically canceled, so future stream events (e.g., stream "done" state) are not sent to this stream subscription
* nothing else happens, so websocket remains disconnected forever, with no future connection attempt, even though autoReconnect is true.

More changes could be put into `_defaultOnStreamError`, but because that is overrideable, it seems better just to keep the reconnection logic in the `onConnectionLost` callback which is called at the end of the stream (which will be reached whenever the websocket connection closes for any reason).

Reply:
Yeah, so we can stay with this design for now, and see if in the next major release we could change some things to make this exception handling more explicative